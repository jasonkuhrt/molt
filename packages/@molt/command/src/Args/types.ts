import type { ParameterSpec } from '../ParameterSpec/index.js'
import type { LineParseError } from './Line/Line.js'

export interface ArgumentReport extends Argument {
  spec: ParameterSpec.Output
  duplicates: Argument[]
  errors: LineParseError[]
}

export interface Argument {
  value: Value
  source: ArgumentSource
}

export type Value =
  | { _tag: 'boolean'; value: boolean; negated: boolean }
  | { _tag: 'number'; value: number }
  | { _tag: 'string'; value: string }

type ArgumentSource = LineSource | EnvironmentSource

interface LineSource {
  _tag: 'line'
  name: string
}

interface EnvironmentSource {
  _tag: 'environment'
  name: string
  namespace: null | string
}
