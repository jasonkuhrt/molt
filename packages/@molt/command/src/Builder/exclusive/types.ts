import type { ParameterSpec } from '../../ParameterSpec/index.js'
import type { RawArgInputs } from '../root/types.js'
import type { State } from '../State.js'
import type { z } from 'zod'

export interface ExclusiveParameterConfiguration {
  schema: ParameterSpec.SomeExclusiveZodType
}

// prettier-ignore
interface Parameter<State extends State.Base, Label extends string> {
  <NameExpression extends string, Configuration extends ExclusiveParameterConfiguration   >(name: State.ValidateNameExpression<State, NameExpression>, configuration: Configuration): BuilderExclusiveInitial<State.AddExclusiveParameter<State, Label, NameExpression, Configuration>, Label>
  <NameExpression extends string, Schema        extends ParameterSpec.SomeExclusiveZodType>(name: State.ValidateNameExpression<State, NameExpression>, schema: Schema              ): BuilderExclusiveInitial<State.AddExclusiveParameter<State, Label, NameExpression, {schema:Schema}>, Label>
}

/**
 * This property is present to support internal functions. It is not intended to be used by you.
 */
export type InternalState<State extends State.Base = State.Base> = {
  /**
   * Used for build time. Type inference functionality.
   */
  typeState: State
  /**
   * Used for runtime.
   */
  input: ParameterSpec.Input.Exclusive
}

// prettier-ignore
export interface BuilderExclusiveInitial<State extends State.Base, Label extends string> {
  _:         InternalState<State>
  parameter: Parameter<State,Label>
  optional:  () => BuilderExclusiveAfterOptional<State.SetExclusiveOptional<State, Label, true>>
  default:  <Tag extends keyof State['ParametersExclusive'][Label]['Parameters']>(tag: Tag, value: z.infer<State['ParametersExclusive'][Label]['Parameters'][Tag]['Schema']>) => BuilderExclusiveAfterDefault<State.SetExclusiveOptional<State,Label,false>>
}

export type BuilderExclusiveAfterOptional<State extends State.Base> = {
  _: InternalState<State>
}

export type BuilderExclusiveAfterDefault<State extends State.Base> = {
  _: InternalState<State>
}

export interface BuilderAfterSettings<Spec extends State.Base> {
  parse: (inputs?: RawArgInputs) => State.ToArgs<Spec>
}

export interface SomeParameter {
  (nameExpression: any, type: ParameterSpec.SomeExclusiveZodType): any // eslint-disable-line
  (nameExpression: any, configuration: ExclusiveParameterConfiguration): any // eslint-disable-line
}

export type SomeBuilderExclusiveInitial = {
  _: any // eslint-disable-line
  parameter: SomeParameter
  optional: any // eslint-disable-line
  default: (tag: any, value: any) => any // eslint-disable-line
}

export type SomeBuilderMutuallyExclusiveAfterOptional = BuilderExclusiveAfterOptional<State.Base>

export type SomeBuilderExclusive = SomeBuilderExclusiveInitial | SomeBuilderMutuallyExclusiveAfterOptional
