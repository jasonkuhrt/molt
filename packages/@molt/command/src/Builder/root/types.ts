import type { OpeningArgs } from '../../OpeningArgs/index.js'
import type { ParameterSpec } from '../../ParameterSpec/index.js'
import type { TTY } from '../../parse/prompt.js'
import type { Settings } from '../../Settings/index.js'
import type {
  BuilderAfterSettings,
  BuilderExclusiveInitial,
  SomeBuilderExclusive,
} from '../exclusive/types.js'
// eslint-disable-next-line
import { State } from '../State.js'

export type Schema = ParameterSpec.SomeBasicType | ParameterSpec.SomeUnionType

export interface ParameterConfiguration {
  schema: Schema
  prompt?: ParameterSpec.Input.Prompt<this['schema']>
}

export type IsHasKey<Obj extends object, Key> = Key extends keyof Obj ? true : false

// prettier-ignore
export type IsPromptEnabledInParameterSettings<P extends ParameterConfiguration> =
  IsHasKey<P,'prompt'>                                      extends false     ? false :
                                                                                IsPromptEnabled<P['prompt']>
// prettier-ignore
export type IsPromptEnabledInCommandSettings<P extends Settings.Input<any>> =
  IsHasKey<P,'prompt'>                                      extends false     ? false :
                                                                                IsPromptEnabled<P['prompt']>

// prettier-ignore
export type IsPromptEnabled<P extends ParameterSpec.Input.Prompt<any>|undefined> =
  P                                               extends undefined ? false :
  P                                               extends false     ? false :
  P                                               extends true      ? true  :
  P                                               extends null      ? false :
  Exclude<P, undefined|boolean|null>['enabled']   extends false     ? false :
                                                                      true

export interface SomeParameterConfig<S extends Schema> {
  schema: S
  prompt?: ParameterSpec.Input.Prompt<S>
}

// prettier-ignore
export interface RootBuilder<State extends State.Base = State.BaseEmpty> {
  s: State
  description                                                                               (this:void, description:string):
    RootBuilder<State>
  parameter<NameExpression extends string, const Configuration extends ParameterConfiguration>    (this:void, name:State.ValidateNameExpression<State,NameExpression>, configuration:Configuration):
    RootBuilder<{
      IsPromptEnabled    : State['IsPromptEnabled'] extends true ? true : IsPromptEnabledInParameterSettings<Configuration>
      ParametersExclusive: State['ParametersExclusive']
      Parameters         : State['Parameters'] & { [_ in NameExpression]: State.CreateParameter<State,NameExpression,Configuration> }
    }>
  parameter<NameExpression extends string, S extends Schema>                                (this:void, name:State.ValidateNameExpression<State,NameExpression>, schema:S):
    RootBuilder<{
      IsPromptEnabled    : State['IsPromptEnabled']
      ParametersExclusive: State['ParametersExclusive']
      Parameters         : State['Parameters'] & { [_ in NameExpression]: State.CreateParameter<State,NameExpression,{schema:S}> }
    }>
  parametersExclusive<Label extends string, BuilderExclusive extends SomeBuilderExclusive>  (this:void, label:Label, ExclusiveBuilderContainer: (builder:BuilderExclusiveInitial<State,Label>) => BuilderExclusive):
    RootBuilder<BuilderExclusive['_']['typeState']>
  settings                                                                                  <S extends Settings.Input<State.ToSchema<State>>>(this:void, newSettings:S):
    BuilderAfterSettings<{
      IsPromptEnabled    : State['IsPromptEnabled'] extends true ? true : IsPromptEnabledInCommandSettings<S>
      ParametersExclusive: State['ParametersExclusive']
      Parameters         : State['Parameters']
    }>
  parse                                                                                     (this:void, inputs?:RawArgInputs):
    State.ToArgs<State>
}

export type RawArgInputs = {
  line?: OpeningArgs.Line.RawInputs
  environment?: OpeningArgs.Environment.RawInputs
  tty?: TTY
}

export type SomeArgsNormalized = Record<string, unknown>
