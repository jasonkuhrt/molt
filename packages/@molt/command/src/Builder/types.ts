import type { FlagSpecExpressionParseResultToPropertyName } from '../helpers.js'
import type { Input } from '../Input/index.js'
import type { ParameterSpec } from '../ParameterSpec/index.js'
import type { Settings } from '../Settings/index.js'
import type { FlagName } from '@molt/types'
import type { Any } from 'ts-toolbelt'
import type { z } from 'zod'

export type RawInputs = {
  line?: Input.Line.RawInputs
  environment?: Input.Environment.RawInputs
}

// prettier-ignore
type ParametersToArguments<ParametersSchema extends z.ZodRawShape> = Any.Compute<{
  [FlagSpecExpression in keyof ParametersSchema as FlagSpecExpressionParseResultToPropertyName<FlagName.Parse<FlagSpecExpression & string>>]:
    z.infer<ParametersSchema[FlagSpecExpression]>
}>

export interface BuilderParameters<ParametersSchema extends z.ZodRawShape> {
  schema: ParametersSchema
  settings: (newSettings: Settings.Input<ParametersSchema>) => BuilderParameters<ParametersSchema>
  parse: (rawInputs?: RawInputs) => ParametersToArguments<ParametersSchema>
}

// eslint-disable-next-line
namespace Spec {
  export type AddExclusiveParameters<
    Spec extends SomeSpec,
    Label extends string,
    Name extends string,
    Schema extends ParameterSpec.SomeZodType
  > = {
    Parameters: Spec['Parameters']
    ParametersExclusive: Spec['ParametersExclusive'] & {
      [k in Label]: {
        [k in Name]: {
          Schema: Schema
          NameParsed: FlagName.Parse<
            Name,
            { usedNames: Values<Spec['Parameters']>['NameUnion']; reservedNames: 'help' | 'h' }
          >
          NameUnion: GetName<
            FlagName.Parse<
              Name,
              { usedNames: Values<Spec['Parameters']>['NameUnion']; reservedNames: 'help' | 'h' }
            >
          >
        }
      }
    }
  }

  export type AddParameter<
    Spec extends SomeSpec,
    Name extends string,
    Schema extends ParameterSpec.SomeZodType
  > = {
    ParametersExclusive: Spec['ParametersExclusive']
    Parameters: Spec['Parameters'] & {
      [k in Name]: {
        Schema: Schema
        NameParsed: FlagName.Parse<
          Name,
          { usedNames: Values<Spec['Parameters']>['NameUnion']; reservedNames: 'help' | 'h' }
        >
        NameUnion: GetName<
          FlagName.Parse<
            Name,
            { usedNames: Values<Spec['Parameters']>['NameUnion']; reservedNames: 'help' | 'h' }
          >
        >
      }
    }
  }

  // prettier-ignore
  export type ToArgs<Spec extends SomeSpec> =
    Any.Compute<
      {
        [Name in keyof Spec['Parameters'] & string as GetCanonicalName<Spec['Parameters'][Name]['NameParsed']>]:
          z.infer<Spec['Parameters'][Name]['Schema']>
      } &
      {
        [Label in keyof Spec['ParametersExclusive'] & string]: Values<{
            [Name in keyof Spec['ParametersExclusive'][Label]]:
              {
                _tag: GetCanonicalName<Spec['ParametersExclusive'][Label][Name]['NameParsed']>
                value: z.infer<Spec['ParametersExclusive'][Label][Name]['Schema']>
              }
        }>
      }
    >
}

// prettier-ignore
export interface Builder<Spec extends SomeSpec> {
  parameter: <Name extends string, Schema extends ParameterSpec.SomeZodType>(
    name: 
    FlagName.Errors.$Is<FlagName.Parse<Name, { usedNames: Values<Spec['Parameters']>['NameUnion'], reservedNames: 'help' | 'h' }>> extends true
      ?                 FlagName.Parse<Name, { usedNames: Values<Spec['Parameters']>['NameUnion'], reservedNames: 'help' | 'h' }>
      : Name,
    schema: Schema
  ) =>
    Builder<Spec.AddParameter<Spec,Name,Schema>>
  parametersExclusive: <Label extends string, BuilderReturned extends SomeBuilderExclusiveParameters>(label: Label, mutuallyExclusiveParametersDefiner: (builder: BuilderExclusiveParameters<Spec, Label>) => BuilderReturned) => Builder<BuilderReturned['_']>
  settings: (newSettings: Settings.Input<SpecToSchema<Spec>>) => BuilderAfterSettings<Spec>
  parse: (inputs?: RawInputs) => Spec.ToArgs<Spec>
}

type SomeBuilderExclusiveParameters =
  | BuilderExclusiveParameters<SomeSpec, string>
  | BuilderMutuallyExclusiveParameterAfterOptional<SomeSpec>

export interface BuilderExclusiveParameters<Spec extends SomeSpec, Label extends string> {
  _: Spec
  parameter: <Name extends string, Schema extends ParameterSpec.SomeZodType>(
    name: FlagName.Errors.$Is<
      FlagName.Parse<
        Name,
        { usedNames: Values<Spec['Parameters']>['NameUnion']; reservedNames: 'help' | 'h' }
      >
    > extends true
      ? FlagName.Parse<
          Name,
          { usedNames: Values<Spec['Parameters']>['NameUnion']; reservedNames: 'help' | 'h' }
        >
      : Name,
    schema: Schema
  ) => BuilderExclusiveParameters<Spec.AddExclusiveParameters<Spec, Label, Name, Schema>, Label>
  optional: () => BuilderMutuallyExclusiveParameterAfterOptional<Spec>
}

export type BuilderMutuallyExclusiveParameterAfterOptional<Spec extends SomeSpec> = {
  _: Spec
}

export interface BuilderAfterSettings<Spec extends SomeSpec> {
  parse: (inputs?: RawInputs) => Spec.ToArgs<Spec>
}

// declare const builder: Builder<{ Parameters: {} }>

// const args = builder
//   .parameter(`a alpha`, z.string())
//   .parameter(`bravo b`, z.number())
//   // .parameter(`b`, z.boolean())
//   // .settings({
//   //   parameters: {
//   //     environment: {

//   //     }
//   //   }
//   // })
//   .parse()

type SomeSpec = {
  ParametersExclusive: {
    [key: string]: {
      [key: string]: {
        NameParsed: FlagName.Types.FlagNames
        NameUnion: string
        Schema: ParameterSpec.SomeZodType
      }
    }
  }
  Parameters: {
    [key: string]: {
      NameParsed: FlagName.Types.FlagNames
      NameUnion: string
      Schema: ParameterSpec.SomeZodType
    }
  }
}

// TODO move to types lib
type GetCanonicalName<Names extends FlagName.Types.FlagNames> = Names['long'] extends string
  ? Names['long']
  : Names['short'] extends string
  ? Names['short']
  : never

// TODO move to types lib
// prettier-ignore
type GetName<Names extends FlagName.Types.SomeParseResult> = Names extends FlagName.Types.FlagNames
  ? (
      | (Names['long'] extends undefined ? never : Names['long'])
      | (Names['short'] extends undefined ? never : Names['short'])
      | Names['aliases']['long'][number]
      | Names['aliases']['short'][number]
    )
  : ''

type Values<T> = T[keyof T]

type SpecToSchema<Spec extends SomeSpec> = {
  [K in keyof Spec['Parameters'] & string as GetCanonicalName<
    Spec['Parameters'][K]['NameParsed']
  >]: Spec['Parameters'][K]['Schema']
}
export type State = {
  settings: Settings.Normalized
  parameterSpecInputs: ParameterSpec.SomeSpecInput
}

export type SomeArgsNormalized = Record<string, unknown>
