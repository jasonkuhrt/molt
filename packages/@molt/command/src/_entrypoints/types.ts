export type { Settings } from '../Settings/index.js'
