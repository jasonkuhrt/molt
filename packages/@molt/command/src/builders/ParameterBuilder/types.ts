import type { $, Objects } from 'hotscript'
import type { TypeBuilder } from '../TypeBuilder/types.js'
import type { PrivateData } from '../../lib/PrivateData/PrivateData.js'
import type { Optionality, OptionalityOptional } from '../../Type/helpers.js'
import type { Type } from '../../Type/index.js'

export type ParameterBuilderWithMinimumState<
  $StateNew extends Partial<State.Base>,
> = ParameterBuilderUpdateState<ParameterBuilder<State.Base>, $StateNew>

export type ParameterBuilderUpdateState<
  $Builder extends ParameterBuilder<State.Base>,
  $StateNew extends Partial<State.Base>,
  // @ts-expect-error fixme
> = ParameterBuilder<PrivateData.UpdateObject<$Builder, $StateNew>>

export type ParameterBuilderUpdateStateProperty<
  $Builder extends ParameterBuilder<State.Base>,
  $P extends keyof PrivateData.Get<$Builder>,
  $V extends PrivateData.Get<$Builder>[$P],
  // @ts-expect-error fixme
> = ParameterBuilder<PrivateData.UpdateProperty<$Builder, $P, $V>>

export type ParameterBuilderInfer<
  $ParameterBuilder extends ParameterBuilderWithStateTypeBuilder,
> = PrivateData.Get<$ParameterBuilder>['optionality']['_tag'] extends 'optional'
  ?
      | Type.Infer<
          PrivateData.Get<
            PrivateData.Get<$ParameterBuilder>['typeBuilder']
          >['type']
        >
      | undefined
  : Type.Infer<
      PrivateData.Get<PrivateData.Get<$ParameterBuilder>['typeBuilder']>['type']
    >

export type ParameterBuilder<$State extends State.Base = State.Base> =
  PrivateData.Set<
    $State,
    {
      name<const $Name extends string>(
        this: void,
        name: $Name,
      ): ParameterBuilder<
        $<
          // Objects.Update<'excludeMethods', Tuples.Append<'name'>>,
          Objects.Update<'name', $Name>,
          $State
        >
      >

      description(
        this: void,
        description: string,
      ): ParameterBuilder<$<Objects.Update<'description', string>, $State>>

      type<$TypeBuilder extends TypeBuilder>(
        this: void,
        type: $TypeBuilder,
      ): ParameterBuilderUpdateStateProperty<
        ParameterBuilder,
        'typeBuilder',
        $TypeBuilder
      >

      optional(
        this: void,
      ): ParameterBuilder<
        $<Objects.Update<'optionality', { _tag: 'optional' }>, $State>
      >

      prompt<$Config extends State.PromptInput>(
        config: $Config,
      ): ParameterBuilder<
        $<
          Objects.Update<
            'prompt',
            {
              enabled: $Config['enabled'] extends boolean
                ? $Config['enabled']
                : true
              when: $Config['when']
            }
          >,
          $State
        >
      >
      prompt(): ParameterBuilder<
        $<Objects.Update<'prompt', { enabled: true; when: object }>, $State>
      >
      prompt<$Enabled extends boolean>(
        enabled: $Enabled,
      ): ParameterBuilder<
        $<Objects.Update<'prompt', { enabled: $Enabled; when: object }>, $State>
      >
    } & (null extends $State['typeBuilder']
      ? {}
      : {
          default(
            this: void,
            value: Type.Infer<
              PrivateData.Get<Exclude<$State['typeBuilder'], null>>['type']
            >,
          ): ParameterBuilder<
            $<Objects.Update<'optionality', { _tag: 'default' }>, $State>
          >
        })
  >

export type ParameterBuilderWithStateTypeBuilder =
  PrivateData.HostUpdateProperty<ParameterBuilder, 'typeBuilder', TypeBuilder>

export namespace State {
  export type PromptInput = {
    enabled?: boolean
    when: object
  }
  export type Prompt = {
    enabled: boolean
    when: object
  }
  export interface Base {
    name: null | string
    description: null | string
    typeBuilder: null | TypeBuilder
    optionality: Optionality
    prompt: null | Prompt
  }
  export interface Initial {
    name: null
    description: null
    typeBuilder: null
    prompt: null
    optionality: OptionalityOptional
  }
  export const initial: Initial = {
    description: null,
    name: null,
    typeBuilder: null,
    prompt: null,
    optionality: { _tag: `optional` },
  }
}
