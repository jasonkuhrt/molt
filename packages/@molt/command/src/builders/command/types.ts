import type { SomeExtension } from '../../extension.js'
import type { HKT } from '../../helpers.js'
import type { Prompter } from '../../lib/Prompter/Prompter.js'
import type { OpeningArgs } from '../../OpeningArgs/index.js'
import type { Prompt } from '../../Parameter/types.js'
import type { Settings } from '../../Settings/index.js'
import type { BuilderExclusiveInitial, SomeBuilderExclusive } from '../exclusive/types.js'
import type { BuilderCommandState } from './state.js'
// todo
// eslint-disable-next-line
import type { Objects, Pipe } from 'hotscript'

export interface ParameterConfiguration<
  $State extends BuilderCommandState.Base = BuilderCommandState.BaseEmpty,
> {
  type: $State['Type']
  prompt?: Prompt<HKT.Call<$State['TypeMapper'], this['type']>>
}

export type IsHasKey<Obj extends object, Key> = Key extends keyof Obj ? true : false

export type IsPromptEnabledInParameterSettings<P extends ParameterConfiguration<any>> = IsHasKey<P, 'prompt'> extends
  false ? false
  : IsPromptEnabled<P['prompt']>

export type IsPromptEnabledInCommandSettings<P extends Settings.Input<any>> = IsHasKey<P, 'prompt'> extends false
  ? false
  : IsPromptEnabled<P['prompt']>

export type IsPromptEnabled<P extends Prompt<any> | undefined> = P extends undefined ? false
  : P extends false ? false
  : P extends true ? true
  : P extends null ? false
  : Exclude<P, undefined | boolean | null>['enabled'] extends false ? false
  : true

export interface CommandBuilder<$State extends BuilderCommandState.Base = BuilderCommandState.BaseEmpty> {
  use<$Extension extends SomeExtension>(extension: $Extension): CommandBuilder<{
    IsPromptEnabled: $State['IsPromptEnabled']
    Parameters: $State['Parameters']
    ParametersExclusive: $State['ParametersExclusive']
    Type: $Extension['types']['type']
    TypeMapper: $Extension['types']['typeMapper']
  }>
  description(this: void, description: string): CommandBuilder<$State>
  parameter<NameExpression extends string, const Configuration extends ParameterConfiguration<$State>>(
    this: void,
    name: BuilderCommandState.ValidateNameExpression<$State, NameExpression>,
    configuration: Configuration,
  ): CommandBuilder<BuilderCommandState.AddParameter<$State, NameExpression, Configuration>>
  parameter<NameExpression extends string, $Type extends $State['Type']>(
    this: void,
    name: BuilderCommandState.ValidateNameExpression<$State, NameExpression>,
    type: $Type,
  ): CommandBuilder<BuilderCommandState.AddParameter<$State, NameExpression, { type: $Type }>>
  parametersExclusive<Label extends string, BuilderExclusive extends SomeBuilderExclusive<$State>>(
    this: void,
    label: Label,
    ExclusiveBuilderContainer: (builder: BuilderExclusiveInitial<$State, Label>) => BuilderExclusive,
  ): CommandBuilder<BuilderExclusive['_']['typeState']>
  settings<S extends Settings.Input<$State>>(
    this: void,
    newSettings: S,
  ): CommandBuilder<
    Pipe<$State, [
      Objects.Update<
        'IsPromptEnabled',
        Objects.Assign<$State['IsPromptEnabled'] extends true ? true : IsPromptEnabledInCommandSettings<S>>
      >,
    ]>
  >
  parse(this: void, inputs?: RawArgInputs): BuilderCommandState.ToArgs<$State>
}

export type RawArgInputs = {
  line?: OpeningArgs.Line.RawInputs
  environment?: OpeningArgs.Environment.RawInputs
  tty?: Prompter
}

export type SomeArgsNormalized = Record<string, unknown>
