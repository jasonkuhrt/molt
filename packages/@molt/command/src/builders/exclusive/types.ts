import type { Type } from '../../Type/index.js'
import type { BuilderCommandState } from '../command/state.js'
import type { BuilderParameterExclusiveState } from './state.js'

export interface ExclusiveParameterConfiguration<$State extends BuilderCommandState.Base> {
  type: $State['Type']
}

// prettier-ignore
interface Parameter<$State extends BuilderCommandState.Base, Label extends string> {
  <NameExpression extends string, Configuration extends ExclusiveParameterConfiguration<$State>>(
    name: BuilderCommandState.ValidateNameExpression<$State, NameExpression>,
    configuration: Configuration,
  ): BuilderExclusiveInitial<
    BuilderCommandState.AddExclusiveParameter<$State, Label, NameExpression, Configuration>,
    Label
  >

  <NameExpression extends string, $Type extends $State['Type']>(
    name: BuilderCommandState.ValidateNameExpression<$State, NameExpression>,
    type: $Type,
  ): // BuilderCommandState.AddExclusiveParameter<$State, Label, NameExpression, { type: $Type }>
  BuilderExclusiveInitial<
    BuilderCommandState.AddExclusiveParameter<$State, Label, NameExpression, { type: $Type }>,
    Label
  >
}

// prettier-ignore
export interface BuilderExclusiveInitial<$State extends BuilderCommandState.Base, Label extends string> {
  _: BuilderParameterExclusiveState<$State>
  parameter: Parameter<$State, Label>
  optional: () => BuilderExclusiveAfterOptional<BuilderCommandState.SetExclusiveOptional<$State, Label, true>>
  default: <Tag extends keyof $State['ParametersExclusive'][Label]['Parameters']>(
    tag: Tag,
    value: Type.Infer<$State['ParametersExclusive'][Label]['Parameters'][Tag]['Type']>,
  ) => BuilderExclusiveAfterDefault<BuilderCommandState.SetExclusiveOptional<$State, Label, false>>
}

export type BuilderExclusiveAfterOptional<$State extends BuilderCommandState.Base> = {
  _: BuilderParameterExclusiveState<$State>
}

export type BuilderExclusiveAfterDefault<$State extends BuilderCommandState.Base> = {
  _: BuilderParameterExclusiveState<$State>
}

export interface SomeParameter<$State extends BuilderCommandState.Base> {
  (nameExpression: any, type: $State['Type']): any // eslint-disable-line
  (nameExpression: any, configuration: ExclusiveParameterConfiguration<$State>): any // eslint-disable-line
}

export type SomeBuilderExclusiveInitial<
  $State extends BuilderCommandState.Base = BuilderCommandState.BaseEmpty,
> = {
  _: any // eslint-disable-line
  parameter: SomeParameter<$State>
  optional: any // eslint-disable-line
  default: (tag: any, value: any) => any // eslint-disable-line
}

export type SomeBuilderMutuallyExclusiveAfterOptional<$State extends BuilderCommandState.Base> =
  BuilderExclusiveAfterOptional<$State>

export type SomeBuilderExclusive<$State extends BuilderCommandState.Base> =
  | SomeBuilderExclusiveInitial<$State>
  | SomeBuilderMutuallyExclusiveAfterOptional<$State>
