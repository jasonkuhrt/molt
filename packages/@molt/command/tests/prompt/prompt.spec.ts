import type { Schema } from '../../src/Builder/root/types.js'
import type { Settings } from '../../src/entrypoints/types.js'
import { Methods } from '../../src/entrypoints/types.js'
import { Command } from '../../src/index.js'
import { s, tryCatch } from '../_/helpers.js'
import { tty } from '../_/mocks/tty.js'
import stripAnsi from 'strip-ansi'
import { describe, expect, it } from 'vitest'

let parameters: Methods.Parameters.InputAsConfig<Schema>
let ttyInputs: string[]
let line: string[]
const settings: Settings.Input = {}

describe(`prompt can be configured at the parameter level`, () => {
  it(`prompt when missing input`, () => {
    parameters = {
      a: { schema: s.min(2), prompt: { when: { result: `rejected`, error: `ErrorMissingArgument` } } },
    }
    ttyInputs = [`foo`]
    line = []
    run()
  })

  it(`prompt when invalid input`, () => {
    parameters = {
      a: { schema: s.min(2), prompt: { when: { result: `rejected`, error: `ErrorInvalidArgument` } } },
    }
    ttyInputs = [`foo`]
    line = [`-a`, `1`]
    run()
  })

  it(`prompt when invalid input OR missing input`, () => {
    parameters = {
      a: {
        schema: s.min(2),
        prompt: { when: { result: `rejected`, error: [`ErrorInvalidArgument`, `ErrorMissingArgument`] } },
      },
    }
    ttyInputs = [`foo`]
    line = [`-a`, `1`]
    run()
  })

  it(`prompt when omitted`, () => {
    parameters = Methods.Parameters.parameters({
      a: {
        schema: s.min(2).optional(),
        prompt: { when: { result: `omitted`, spec: { optionality: [`default`, `optional`] } } },
      },
    })
    ttyInputs = [`foo`]
    line = []
    run()
  })

  it(`static error to match on omitted event on required parameter`, () => {
    // @ts-expect-error not available
    Command.parameters({ a: { schema: s, prompt: { when: { result: `omitted` } } } })
    // Is fine, because parameter is optional.
    Command.parameters({ a: { schema: s.optional(), prompt: { when: { result: `omitted` } } } })
  })
})

const run = () => {
  tty.mock.input.add(ttyInputs)
  const args = tryCatch(() =>
    Command.parameters(parameters)
      .settings({ onError: `throw`, helpOnError: false, ...settings })
      .parse({ line, tty: tty.interface }),
  )
  expect(args).toMatchSnapshot(`args`)
  expect(tty.history.all).toMatchSnapshot(`tty`)
  expect(tty.history.all.map((_) => stripAnsi(_))).toMatchSnapshot(`tty strip ansi`)
}

describe(`types`, () => {
  it(`rejected name must be a valid string literal`, () => {
    parameters = {
      a: {
        schema: s.min(2),
        prompt: {
          when: {
            // @ts-expect-error invalid type
            rejected: {
              name: `bad`,
            },
          },
        },
      },
    }
  })
})
