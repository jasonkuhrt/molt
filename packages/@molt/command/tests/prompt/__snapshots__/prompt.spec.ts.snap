// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`a parameter with prompt enabled given an invalid value at line does not prompt > args 1`] = `
[Error: Invalid value for a:
Value is too short.]
`;

exports[`a parameter with prompt enabled given an invalid value at line does not prompt > tty 1`] = `[]`;

exports[`a required parameter can be given at prompt > args 1`] = `
{
  "a": "foo",
  "help": false,
}
`;

exports[`a required parameter can be given at prompt > tty 1`] = `
[
  "Please give argument for parameter \\"a\\"",
  "foo",
]
`;

exports[`a required parameter given by flag is not asked for in prompt > args 1`] = `
{
  "a": "foo",
  "help": false,
}
`;

exports[`a required parameter given by flag is not asked for in prompt > tty 1`] = `[]`;

exports[`a required parameter opted out of prompt is not prompted for > args 1`] = `[Error: Missing argument for flag "a".]`;

exports[`a required parameter opted out of prompt is not prompted for > tty 1`] = `[]`;

exports[`an optional parameter can be skipped at prompt > args 1`] = `
{
  "a": undefined,
  "help": false,
}
`;

exports[`an optional parameter can be skipped at prompt > tty 1`] = `[]`;

exports[`by default parameter prompt is disabled > args 1`] = `[Error: Missing argument for flag "a".]`;

exports[`by default parameter prompt is disabled > tty 1`] = `[]`;

exports[`prompt input is validated > args 1`] = `
{
  "a": "xx",
  "help": false,
}
`;

exports[`prompt input is validated > tty 1`] = `
[
  "Please give argument for parameter \\"a\\"",
  "x",
  "Invalid value: Value is too short.",
  "xx",
]
`;

exports[`when there is no tty then prompt is skipped > args 1`] = `[Error: Missing argument for flag "a".]`;

exports[`when there is no tty then prompt is skipped > tty 1`] = `[]`;
